src/qibolab/instruments/emulator/pulse_simulator.py

```diff
@@ -0,0 +1,711 @@
+"""Pulse simulator module for running quantum dynamics simulation of model of
+device."""
+
+import operator
+from dataclasses import dataclass
+from typing import Dict, List, Optional, Union
+
+import numpy as np
+from qibo.config import log
+
+from qibolab import AcquisitionType, AveragingMode, ExecutionParameters
+from qibolab.instruments.abstract import Controller
+from qibolab.instruments.emulator.backends.generic import make_comp_basis
+from qibolab.instruments.emulator.backends.qutip_backend import QutipSimulator
+from qibolab.instruments.port import Port
+from qibolab.platform import Coupler, Qubit
+from qibolab.pulses import PulseSequence, ReadoutPulse
+from qibolab.qubits import QubitId
+from qibolab.result import IntegratedResults, SampleResults
+from qibolab.sweeper import Parameter, Sweeper, SweeperType
+
+
+@dataclass
+class DummyPort(Port):
+    """Placeholder.
+
+    Copied over from dummy, may not be required.
+    """
+
+    name: str
+    offset: float = 0.0
+    lo_frequency: int = 0
+    lo_power: int = 0
+    gain: int = 0
+    attenuation: int = 0
+    power_range: int = 0
+    filters: Optional[dict] = None
+
+
+class PulseSimulator(Controller):
+    """Runs quantum dynamics simulation of model of device.
+
+    Interfaces with Qibolab. Useful for testing code without requiring
+    access to hardware.
+    """
+
+    PortType = DummyPort  # Placeholder. Copied over from dummy, may not be required.
+
+    def __init__(
+        self,
+        simulation_config: dict,
+        model_config: dict,
+        sim_opts: Optional[None] = None,
+    ):
+        """Set emulator configuration.
+
+        Args:
+            simulation_config (dict): Simulation configuration dictionary.
+            model_config (dict): Model configuration dictionary.
+            sim_opts (optional): Simulation backend specific object specifying simulation options.
+        """
+        self.simulation_config = simulation_config
+        self.model_config = model_config
+        self.sim_opts = sim_opts
```

Comment:
Since these are read and consumed in `update()`, after which you shouldn't need them any longer, avoid assigning them to `self`.

The minimal improvement would be to just pass them as parameters to `self.update()` call below, instead of communicating through instance attributes.
(long term, I believe there should not be any `.update()` method, but I'll comment on this below)

Reply:
Fixed in 7f4b9230dd3d7291758bb3681abb75570ec34bf6